File: frontend/src/types/task.ts
// frontend/src/types/task.ts - Tipos TypeScript para Tasks

// Enums para Task
export type TaskStatus = 'pending' | 'in_progress' | 'completed' | 'cancelled'

export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent'

// Interface base para Task
export interface TaskBase {
  title: string
  description: string | null
  status: TaskStatus
  priority: TaskPriority
  due_date: string | null
}

// Interface para crear Task
export interface TaskCreate extends TaskBase {
  contact_id?: number | null
  deal_id?: number | null
}

// Interface para actualizar Task
export interface TaskUpdate extends Partial<TaskBase> {
  contact_id?: number | null
  deal_id?: number | null
}

// Interface para Task completa (respuesta del servidor)
export interface TaskOut extends TaskBase {
  id: number
  assigned_to_id: number
  contact_id: number | null
  deal_id: number | null
  completed_at: string | null
  created_at: string
  updated_at: string
}

// Interface para Task con información adicional
export interface TaskWithDetails extends TaskOut {
  contact_name: string | null
  deal_title: string | null
  assigned_to_name: string | null
}

// Interface para respuesta paginada
export interface TasksPaginatedResponse {
  items: TaskOut[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// Interface para estadísticas de tasks
export interface TaskStats {
  total_tasks: number
  pending_tasks: number
  in_progress_tasks: number
  completed_tasks: number
  cancelled_tasks: number
  tasks_by_priority: Record<TaskPriority, number>
  tasks_by_status: Record<TaskStatus, number>
  overdue_tasks: number
  due_today: number
  due_this_week: number
  completion_rate: number
}

// Labels para los estados
export const StatusLabels: Record<TaskStatus, string> = {
  pending: 'Pendiente',
  in_progress: 'En Progreso',
  completed: 'Completada',
  cancelled: 'Cancelada',
}

// Labels para prioridades
export const PriorityLabels: Record<TaskPriority, string> = {
  low: 'Baja',
  medium: 'Media',
  high: 'Alta',
  urgent: 'Urgente',
}

// Opciones para selects
export const STATUS_OPTIONS = Object.entries(StatusLabels).map(([value, label]) => ({
  value: value as TaskStatus,
  label,
}))

export const PRIORITY_OPTIONS = Object.entries(PriorityLabels).map(([value, label]) => ({
  value: value as TaskPriority,
  label,
}))

// Colores para estados (para UI)
export const StatusColors: Record<TaskStatus, string> = {
  pending: 'gray',
  in_progress: 'blue',
  completed: 'green',
  cancelled: 'red',
}

// Colores para prioridades
export const PriorityColors: Record<TaskPriority, string> = {
  low: 'gray',
  medium: 'blue',
  high: 'orange',
  urgent: 'red',
}

// Iconos para estados
export const StatusIcons: Record<TaskStatus, string> = {
  pending: 'CircleDashed',
  in_progress: 'Loader2',
  completed: 'CheckCircle2',
  cancelled: 'XCircle',
}

// Iconos para prioridades
export const PriorityIcons: Record<TaskPriority, string> = {
  low: 'ArrowDown',
  medium: 'Minus',
  high: 'ArrowUp',
  urgent: 'AlertTriangle',
}

File: backend/app/schemas/task.py
# backend/app/schemas/task.py - Schemas completos para Tasks

from pydantic import BaseModel, Field, field_validator
from datetime import datetime
from typing import Optional, List

from app.models.task import TaskStatus, TaskPriority


# Schema base para Task
class TaskBase(BaseModel):
    title: str = Field(..., min_length=2, max_length=300, description="Título de la tarea")
    description: Optional[str] = Field(None, description="Descripción detallada")
    status: TaskStatus = Field(default=TaskStatus.PENDING, description="Estado de la tarea")
    priority: TaskPriority = Field(default=TaskPriority.MEDIUM, description="Prioridad")
    due_date: Optional[datetime] = Field(None, description="Fecha de vencimiento")

    @field_validator("status", "priority", mode="before")
    @classmethod
    def normalize_enums(cls, v):
        """Normaliza enums a mayúsculas"""
        if isinstance(v, str):
            return v.upper()
        return v


# Schema para crear Task
class TaskCreate(TaskBase):
    contact_id: Optional[int] = Field(None, description="ID del contacto asociado")
    deal_id: Optional[int] = Field(None, description="ID del deal asociado")

    @field_validator('title')
    @classmethod
    def validate_title(cls, v):
        if not v or not v.strip():
            raise ValueError('El título es obligatorio')
        return v.strip()


# Schema para actualizar Task
class TaskUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=2, max_length=300)
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    due_date: Optional[datetime] = None
    contact_id: Optional[int] = None
    deal_id: Optional[int] = None

    @field_validator("status", "priority", mode="before")
    @classmethod
    def normalize_enums(cls, v):
        """Normaliza enums a mayúsculas"""
        if isinstance(v, str):
            return v.upper()
        return v


# Schema para respuesta de Task
class TaskOut(TaskBase):
    id: int
    assigned_to_id: int
    contact_id: Optional[int] = None
    deal_id: Optional[int] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Schema para Task con información adicional
class TaskWithDetails(TaskOut):
    contact_name: Optional[str] = None
    deal_title: Optional[str] = None
    assigned_to_name: Optional[str] = None


# Schema para paginación de Tasks
class TasksPaginatedResponse(BaseModel):
    items: List[TaskOut]
    total: int
    page: int
    page_size: int
    total_pages: int


# Schema para estadísticas de Tasks
class TaskStats(BaseModel):
    # Totales
    total_tasks: int
    pending_tasks: int
    in_progress_tasks: int
    completed_tasks: int
    cancelled_tasks: int

    # Por prioridad
    tasks_by_priority: dict  # {priority: count}

    # Por estado
    tasks_by_status: dict  # {status: count}

    # Vencimiento
    overdue_tasks: int
    due_today: int
    due_this_week: int

    # Porcentajes
    completion_rate: float  # Porcentaje de tareas completadas
